### DISCUSSION TITLE
Can I use Vivaldi/Brave source code in Bromite?

### DISCUSSION THREAD

**Discussion by uazo** [bromite/bromite#1504] [Q&A]
I am not an expert on licensing at all, could someone tell me if it is legal and in what form it is possible to use code already present in Vivaldi or Brave?
do i have to warn them? do i have to have their authorization?
many thanks in advance.

**Comment by uazo** [comment]
in part I answered myself, vivaldi is [closed-source](https://vivaldi.com/blog/technology/why-isnt-vivaldi-browser-open-source). most android code in fact is not available, too bad

**Reply by csagan5** [reply]
It depends from the license; if you find any code that is compatible with the licenses used in Chromium, you can add it. You do not need any authorisation from anyone, as far as you follow the license.

**Reply by uazo** [reply]
ok, practical example:
brave is under the MPLv2.0 license which, if I have not misunderstood, obliges the use of derivative products under the same license, which is therefore incompatible with the GNUv3 of bromite, correct?

section 3.3 of the mpl speaks of the possibility of use in the case of "Larger Work" but I do not understand if they intend only the original without modification... and, above all, is a modification, for example, a change of a variable name?

so, when you do not copy the code directly, but derive it from it, even just by changing the names of the functions for example, is it possible?

sorry for these questions if I waste your time, for me it is still truly a jungle, and I have not yet found something that explains it to me well.

**Reply by csagan5** [reply]
> brave is under the MPLv2.0 license which, if I have not misunderstood, obliges the use of derivative products under the same license, which is therefore incompatible with the GNUv3 of bromite, correct?

Perhaps https://www.mozilla.org/en-US/MPL/2.0/FAQ/ has some answers.

> sorry for these questions if I waste your time, for me it is still truly a jungle, and I have not yet found something that explains it to me well.

No worries; MPL or GPL only apply to the code which is fully created for Brave/Bromite, and to the patch files themselves in the case of Bromite. If existing code is changed, the original license still applies. This is how the Chromium codebase works, since it has a lot of different licenses.

**Reply by uazo** [reply]
thank you very much for the information.
if you agree, let's do that, since I've seen some interesting features in brave and vivaldi, I'll show you the source so you tell me how to behave.

for now I have seen interesting:
- support for external downloaders (vivaldi)
- fingerprint mitigations (brave)

I haven't finished analyzing their code yet, maybe I find something else too.

**Reply by zaggertu** [reply]
@uazo Please take a look at brave autoplay blocking method aswell, it has a per site toggle (like images, ads, cookies, etc in bromite) and it doesn't break webrtc like the patch that you made for bromite when autoplay is allowed for that site, it's been made by @wchen342 .

**Reply by uazo** [reply]
> it doesn't break webrtc like the patch that you made

it is not a bug, it is intended.
https://github.com/bromite/bromite/issues/791#issuecomment-724883548
https://github.com/bromite/bromite/issues/804

**Reply by csagan5** [reply]
> for now I have seen interesting:
> 
> * support for external downloaders (vivaldi)
> * fingerprint mitigations (brave)

For the first one I don't think it is related to privacy at all, so if it involves complex patches we should not venture into that.
As for fingerprinting mitigations: you could also re-implement the functionality, but it is a complex topic. Bromite does not aim to "mix with the crowd", since it cannot; the mitigations we have in place are about hiding or faking data that the browser would normally expose.

**Reply by uazo** [reply]
> For the first one I don't think it is related to privacy at all, so if it involves complex patches we should not venture into that.

certainly, we are at an idea level, I don't even know if I would be able to develop it.

> As for fingerprinting mitigations: you could also re-implement the functionality, but it is a complex topic.

whatever may ever come out will be subjected to your careful review, at the most you can not accept it.

**Reply by peternrdstrm** [reply]
> As for fingerprinting mitigations: you could also re-implement the functionality, but it is a complex topic. Bromite does not aim to "mix with the crowd", since it cannot; the mitigations we have in place are about hiding or faking data that the browser would normally expose.

Sorry to interupt this conversation, I really don't mean to bloat it. As far as I can understand, Brave's fingerprinting protections are also randomization based. I might be wrong, but they use canvas, audio and webgl randomization on standard, and disable webgl completely on strict. 
As an addition, I might suggest font fingerprinting too. 
I barely have knowleadge of coding, so I'm no expert, but that's how I understood it.
